File: exercises/python/02_re_to_nfa/re_to_nfa.py
"""
================================================================================
Exercise 2: Regular Expression to NFA (Thompson's Construction)
================================================================================

CONCEPT:
    Thompson's construction is a systematic method for converting any regular
    expression into an equivalent Non-deterministic Finite Automaton (NFA).
    The resulting NFA has exactly one start state and one accept state, uses
    epsilon-transitions liberally, and each state has at most two outgoing
    transitions.  These properties make the NFA easy to compose recursively.

ALGORITHM:
    1. First, insert explicit concatenation operators ('.') between adjacent
       symbols where concatenation is implied (e.g., "ab" becomes "a.b").
    2. Convert the infix expression to postfix using the shunting-yard
       algorithm, respecting operator precedence: '*' > '.' > '|'.
    3. Walk the postfix expression left to right.  For each symbol:
       - Literal character 'c': push a two-state NFA  start --c--> accept.
       - Concatenation '.':  pop two NFAs, connect first's accept to
         second's start via epsilon.
       - Union '|':  pop two NFAs, create new start/accept with epsilon
         branches to both.
       - Kleene star '*':  pop one NFA, create new start/accept with
         epsilon loops.
    4. The single NFA remaining on the stack is the result.

COMPLEXITY:
    Time:  O(n) where n is the length of the regular expression -- each
           character is processed a constant number of times.
    Space: O(n) states are created (at most 2 per character in the regex).

EXAMPLE:
    Input regex:  (a|b)*abb
    Output: An NFA with epsilon transitions printed as a transition table.
================================================================================
"""

# ---------------------------------------------------------------------------
# NFA building blocks
# ---------------------------------------------------------------------------

EPSILON = "eps"  # label for epsilon transitions

class State:
    """A single NFA state."""
    _counter = 0

    def __init__(self):
        State._counter += 1
        self.id = State._counter
        self.transitions = {}   # char -> list of State

    def add_transition(self, symbol, target):
        self.transitions.setdefault(symbol, []).append(target)

    def __repr__(self):
        return f"q{self.id}"


class NFA:
    """An NFA fragment with a single start state and a single accept state."""
    def __init__(self, start, accept):
        self.start = start
        self.accept = accept


def reset_state_counter():
    State._counter = 0


# ---------------------------------------------------------------------------
# Thompson's construction primitives
# ---------------------------------------------------------------------------
def nfa_for_char(ch):
    """Create a basic NFA for a single character: start --ch--> accept."""
    start = State()
    accept = State()
    start.add_transition(ch, accept)
    return NFA(start, accept)


def nfa_concat(first, second):
    """Concatenation: connect first.accept to second.start via epsilon."""
    first.accept.add_transition(EPSILON, second.start)
    return NFA(first.start, second.accept)


def nfa_union(upper, lower):
    """Union (|): new start branches to both; both accepts merge to new accept."""
    start = State()
    accept = State()
    start.add_transition(EPSILON, upper.start)
    start.add_transition(EPSILON, lower.start)
    upper.accept.add_transition(EPSILON, accept)
    lower.accept.add_transition(EPSILON, accept)
    return NFA(start, accept)


def nfa_star(inner):
    """Kleene star (*): zero or more repetitions."""
    start = State()
    accept = State()
    start.add_transition(EPSILON, inner.start)
    start.add_transition(EPSILON, accept)         # zero occurrences
    inner.accept.add_transition(EPSILON, inner.start)  # repeat
    inner.accept.add_transition(EPSILON, accept)       # exit loop
    return NFA(start, accept)


# ---------------------------------------------------------------------------
# Regex preprocessing: insert explicit concatenation operator '.'
# ---------------------------------------------------------------------------
def insert_concat_operator(regex):
    """
    Insert '.' where concatenation is implied.
    Concatenation is implied between:
      - two literals:          a b   -> a . b
      - literal and '(':      a (   -> a . (
      - ')' and literal:      ) a   -> ) . a
      - ')' and '(':          ) (   -> ) . (
      - '*' and literal:      * a   -> * . a
      - '*' and '(':          * (   -> * . (
    """
    result = []
    for i, ch in enumerate(regex):
        result.append(ch)
        if i + 1 < len(regex):
            next_ch = regex[i + 1]
            if (ch not in ('(', '|', '.') and
                next_ch not in (')', '|', '*', '.')):
                result.append('.')
    return ''.join(result)


# ---------------------------------------------------------------------------
# Shunting-yard: infix to postfix
# ---------------------------------------------------------------------------
PRECEDENCE = {'|': 1, '.': 2, '*': 3}

def infix_to_postfix(infix):
    """Convert infix regex (with explicit '.') to postfix."""
    output = []
    stack = []
    for ch in infix:
        if ch == '(':
            stack.append(ch)
        elif ch == ')':
            while stack and stack[-1] != '(':
                output.append(stack.pop())
            if stack:
                stack.pop()  # remove '('
        elif ch in PRECEDENCE:
            while (stack and stack[-1] != '(' and
                   stack[-1] in PRECEDENCE and
                   PRECEDENCE[stack[-1]] >= PRECEDENCE[ch]):
                output.append(stack.pop())
            stack.append(ch)
        else:
            output.append(ch)  # literal
    while stack:
        output.append(stack.pop())
    return ''.join(output)


# ---------------------------------------------------------------------------
# Build NFA from postfix regex
# ---------------------------------------------------------------------------
def build_nfa(postfix):
    """Evaluate postfix regex using a stack of NFA fragments."""
    nfa_stack = []
    for ch in postfix:
        if ch == '.':
            second = nfa_stack.pop()
            first = nfa_stack.pop()
            nfa_stack.append(nfa_concat(first, second))
        elif ch == '|':
            lower = nfa_stack.pop()
            upper = nfa_stack.pop()
            nfa_stack.append(nfa_union(upper, lower))
        elif ch == '*':
            inner = nfa_stack.pop()
            nfa_stack.append(nfa_star(inner))
        else:
            nfa_stack.append(nfa_for_char(ch))
    return nfa_stack.pop()


# ---------------------------------------------------------------------------
# Collect all states and print the NFA
# ---------------------------------------------------------------------------
def collect_states(nfa):
    """BFS/DFS to collect all reachable states from nfa.start."""
    visited = set()
    queue = [nfa.start]
    states = []
    while queue:
        state = queue.pop(0)
        if state.id in visited:
            continue
        visited.add(state.id)
        states.append(state)
        for targets in state.transitions.values():
            for target in targets:
                if target.id not in visited:
                    queue.append(target)
    states.sort(key=lambda s: s.id)
    return states


def get_alphabet(states):
    """Return the set of input symbols (excluding epsilon)."""
    alphabet = set()
    for state in states:
        for sym in state.transitions:
            if sym != EPSILON:
                alphabet.add(sym)
    return sorted(alphabet)


def print_nfa(nfa):
    """Print the NFA transition table."""
    states = collect_states(nfa)
    alphabet = get_alphabet(states)
    all_symbols = [EPSILON] + alphabet

    # column widths
    state_width = max(6, max(len(str(s)) for s in states) + 4)
    col_width = 14

    # header
    header = f"{'STATE':<{state_width}}"
    for sym in all_symbols:
        header += f" | {sym:^{col_width}}"
    sep = "-" * len(header)

    print(sep)
    print(header)
    print(sep)

    for state in states:
        marker = ""
        if state.id == nfa.start.id:
            marker = "->"
        if state.id == nfa.accept.id:
            marker += "*"
        label = f"{marker}{state}"

        row = f"{label:<{state_width}}"
        for sym in all_symbols:
            targets = state.transitions.get(sym, [])
            cell = ", ".join(str(t) for t in sorted(targets, key=lambda s: s.id))
            if not cell:
                cell = "-"
            row += f" | {cell:^{col_width}}"
        print(row)

    print(sep)
    print(f"\nStart state : {nfa.start}")
    print(f"Accept state: {nfa.accept}")
    print(f"Total states: {len(states)}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main():
    print("=" * 70)
    print("  RE -> NFA (Thompson's Construction)  --  Exercise 2")
    print("=" * 70)

    default_regex = "(a|b)*abb"

    print(f"\nDefault regex: {default_regex}")
    user = input("Enter a regex (or press Enter for default): ").strip()
    regex = user if user else default_regex

    reset_state_counter()

    print(f"\n1. Original regex      : {regex}")

    with_concat = insert_concat_operator(regex)
    print(f"2. With explicit concat: {with_concat}")

    postfix = infix_to_postfix(with_concat)
    print(f"3. Postfix form        : {postfix}")

    nfa = build_nfa(postfix)

    print(f"\n--- NFA Transition Table ---\n")
    print_nfa(nfa)


if __name__ == "__main__":
    main()

File: exercises/c/02_re_to_nfa/re_to_nfa.c
/*
================================================================================
Exercise 2: Regular Expression to NFA (Thompson's Construction) -- C
================================================================================

CONCEPT:
    Thompson's construction converts a regular expression into an equivalent
    Non-deterministic Finite Automaton (NFA).  Each NFA fragment has exactly
    one start state and one accept state.  The construction is recursive:
    small NFAs for individual characters are composed using concatenation,
    union, and Kleene-star operations.

ALGORITHM:
    1. Insert explicit concatenation operators ('.') where implied.
    2. Convert the augmented infix regex to postfix using the shunting-yard
       algorithm with precedence:  * (3)  >  . (2)  >  | (1).
    3. Evaluate the postfix expression with a stack of NFA fragments:
       - Literal 'c'  : push a two-state fragment  start --c--> accept.
       - '.'          : pop two, link first.accept --eps--> second.start.
       - '|'          : pop two, new start with eps to both starts,
                        both accepts eps to new accept.
       - '*'          : pop one, add eps loop and bypass.
    4. The remaining fragment on the stack is the complete NFA.

COMPLEXITY:
    Time:  O(n) -- linear in the length of the regex.
    Space: O(n) -- at most 2*n states.

EXAMPLE:
    Regex: (a|b)*abb   ->  NFA with ~11 states.
================================================================================
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* ---- constants ---------------------------------------------------------- */
#define MAX_STATES      128
#define MAX_TRANSITIONS 4     /* per state per symbol */
#define MAX_SYMBOLS     32
#define MAX_REGEX_LEN   256
#define EPSILON         '\0'  /* we represent epsilon internally as '\0' */

/* ---- state -------------------------------------------------------------- */
typedef struct {
    int   targets[MAX_SYMBOLS][MAX_TRANSITIONS];  /* targets[sym_index][i] */
    int   target_count[MAX_SYMBOLS];
} StateRow;

/* ---- global NFA storage ------------------------------------------------- */
static StateRow nfa_table[MAX_STATES];
static int      state_count = 0;

/* symbol map: index 0 = epsilon, 1.. = input characters */
static char     sym_chars[MAX_SYMBOLS];
static int      sym_count = 0;

static int get_sym_index(char ch) {
    for (int i = 0; i < sym_count; i++)
        if (sym_chars[i] == ch) return i;
    sym_chars[sym_count] = ch;
    return sym_count++;
}

static int new_state(void) {
    int id = state_count++;
    for (int s = 0; s < MAX_SYMBOLS; s++)
        nfa_table[id].target_count[s] = 0;
    return id;
}

static void add_transition(int from, char symbol, int to) {
    int si = get_sym_index(symbol);
    int *cnt = &nfa_table[from].target_count[si];
    if (*cnt < MAX_TRANSITIONS)
        nfa_table[from].targets[si][(*cnt)++] = to;
}

/* ---- NFA fragment ------------------------------------------------------- */
typedef struct {
    int start;
    int accept;
} Fragment;

/* ---- fragment stack ----------------------------------------------------- */
static Fragment frag_stack[MAX_STATES];
static int      frag_top = -1;

static void frag_push(Fragment f)  { frag_stack[++frag_top] = f; }
static Fragment frag_pop(void)     { return frag_stack[frag_top--]; }

/* ---- Thompson primitives ------------------------------------------------ */
static Fragment nfa_for_char(char ch) {
    int s = new_state();
    int a = new_state();
    add_transition(s, ch, a);
    return (Fragment){s, a};
}

static Fragment nfa_concat(Fragment first, Fragment second) {
    add_transition(first.accept, EPSILON, second.start);
    return (Fragment){first.start, second.accept};
}

static Fragment nfa_union(Fragment upper, Fragment lower) {
    int s = new_state();
    int a = new_state();
    add_transition(s, EPSILON, upper.start);
    add_transition(s, EPSILON, lower.start);
    add_transition(upper.accept, EPSILON, a);
    add_transition(lower.accept, EPSILON, a);
    return (Fragment){s, a};
}

static Fragment nfa_star(Fragment inner) {
    int s = new_state();
    int a = new_state();
    add_transition(s, EPSILON, inner.start);
    add_transition(s, EPSILON, a);
    add_transition(inner.accept, EPSILON, inner.start);
    add_transition(inner.accept, EPSILON, a);
    return (Fragment){s, a};
}

/* ---- insert explicit concatenation '.' ---------------------------------- */
static void insert_concat(const char *regex, char *out) {
    int len = (int)strlen(regex);
    int oi = 0;
    for (int i = 0; i < len; i++) {
        out[oi++] = regex[i];
        if (i + 1 < len) {
            char curr = regex[i];
            char next = regex[i + 1];
            if (curr != '(' && curr != '|' && curr != '.' &&
                next != ')' && next != '|' && next != '*' && next != '.') {
                out[oi++] = '.';
            }
        }
    }
    out[oi] = '\0';
}

/* ---- shunting-yard: infix -> postfix ------------------------------------ */
static int precedence(char op) {
    if (op == '*') return 3;
    if (op == '.') return 2;
    if (op == '|') return 1;
    return 0;
}

static void to_postfix(const char *infix, char *postfix) {
    char stack[MAX_REGEX_LEN];
    int  sp = -1;
    int  oi = 0;
    int  len = (int)strlen(infix);

    for (int i = 0; i < len; i++) {
        char ch = infix[i];
        if (ch == '(') {
            stack[++sp] = ch;
        } else if (ch == ')') {
            while (sp >= 0 && stack[sp] != '(')
                postfix[oi++] = stack[sp--];
            if (sp >= 0) sp--;  /* pop '(' */
        } else if (ch == '|' || ch == '.' || ch == '*') {
            while (sp >= 0 && stack[sp] != '(' &&
                   precedence(stack[sp]) >= precedence(ch))
                postfix[oi++] = stack[sp--];
            stack[++sp] = ch;
        } else {
            postfix[oi++] = ch;  /* literal */
        }
    }
    while (sp >= 0) postfix[oi++] = stack[sp--];
    postfix[oi] = '\0';
}

/* ---- build NFA from postfix --------------------------------------------- */
static Fragment build_nfa(const char *postfix) {
    frag_top = -1;
    int len = (int)strlen(postfix);
    for (int i = 0; i < len; i++) {
        char ch = postfix[i];
        if (ch == '.') {
            Fragment second = frag_pop();
            Fragment first  = frag_pop();
            frag_push(nfa_concat(first, second));
        } else if (ch == '|') {
            Fragment lower = frag_pop();
            Fragment upper = frag_pop();
            frag_push(nfa_union(upper, lower));
        } else if (ch == '*') {
            Fragment inner = frag_pop();
            frag_push(nfa_star(inner));
        } else {
            frag_push(nfa_for_char(ch));
        }
    }
    return frag_pop();
}

/* ---- print NFA ---------------------------------------------------------- */
static void print_nfa(Fragment nfa_frag) {
    /* build ordered symbol list: epsilon first, then alphabetical */
    int eps_idx = -1;
    for (int i = 0; i < sym_count; i++)
        if (sym_chars[i] == EPSILON) { eps_idx = i; break; }

    int ordered[MAX_SYMBOLS];
    int oc = 0;
    if (eps_idx >= 0) ordered[oc++] = eps_idx;
    for (int i = 0; i < sym_count; i++)
        if (i != eps_idx) ordered[oc++] = i;

    /* header */
    printf("%-10s", "STATE");
    for (int i = 0; i < oc; i++) {
        char label[8];
        if (sym_chars[ordered[i]] == EPSILON)
            snprintf(label, sizeof(label), "eps");
        else
            snprintf(label, sizeof(label), "%c", sym_chars[ordered[i]]);
        printf("| %-14s", label);
    }
    printf("\n");

    /* separator */
    for (int i = 0; i < 10 + oc * 16; i++) putchar('-');
    printf("\n");

    /* rows */
    for (int st = 0; st < state_count; st++) {
        char prefix[16] = "";
        if (st == nfa_frag.start)  strcat(prefix, "->");
        if (st == nfa_frag.accept) strcat(prefix, "*");
        char label[24];
        snprintf(label, sizeof(label), "%sq%d", prefix, st);
        printf("%-10s", label);

        for (int i = 0; i < oc; i++) {
            int si = ordered[i];
            int cnt = nfa_table[st].target_count[si];
            if (cnt == 0) {
                printf("| %-14s", "-");
            } else {
                char buf[64] = "";
                for (int k = 0; k < cnt; k++) {
                    char tmp[16];
                    snprintf(tmp, sizeof(tmp), "%sq%d", k > 0 ? ", " : "",
                             nfa_table[st].targets[si][k]);
                    strcat(buf, tmp);
                }
                printf("| %-14s", buf);
            }
        }
        printf("\n");
    }

    for (int i = 0; i < 10 + oc * 16; i++) putchar('-');
    printf("\n");

    printf("\nStart state : q%d\n", nfa_frag.start);
    printf("Accept state: q%d\n", nfa_frag.accept);
    printf("Total states: %d\n", state_count);
}

/* ---- main --------------------------------------------------------------- */
int main(void)
{
    printf("======================================================================\n");
    printf("  RE -> NFA (Thompson's Construction)  --  Exercise 2\n");
    printf("======================================================================\n\n");

    const char *default_regex = "(a|b)*abb";

    printf("Default regex: %s\n", default_regex);
    printf("Enter a regex (or press Enter for default): ");

    char input_buf[MAX_REGEX_LEN];
    if (fgets(input_buf, sizeof(input_buf), stdin) == NULL)
        input_buf[0] = '\n';
    /* strip newline */
    input_buf[strcspn(input_buf, "\n")] = '\0';

    const char *regex = (strlen(input_buf) > 0) ? input_buf : default_regex;

    /* initialize epsilon symbol */
    get_sym_index(EPSILON);

    char with_concat[MAX_REGEX_LEN];
    insert_concat(regex, with_concat);

    char postfix[MAX_REGEX_LEN];
    to_postfix(with_concat, postfix);

    printf("\n1. Original regex      : %s\n", regex);
    printf("2. With explicit concat: %s\n", with_concat);
    printf("3. Postfix form        : %s\n", postfix);

    Fragment result = build_nfa(postfix);

    printf("\n--- NFA Transition Table ---\n\n");
    print_nfa(result);

    return 0;
}
